istioctl/pkg/rbac/upgrade.go

```diff
@@ -0,0 +1,296 @@
+		if subject.Group != "" {
+			subject.Groups = []string{subject.Group}
+			subject.Group = ""
+		}
+		if bindingSpec.RoleRef != nil {
+			bindingSpec.Role = bindingSpec.RoleRef.Name
+			bindingSpec.RoleRef = nil
+		}
+	}
+	authzPolicy := rbacproto.AuthorizationPolicy{}
+	authzPolicy.Allow = []*rbacproto.ServiceRoleBinding{bindingSpec}
+	workloadLabels := WorkloadLabels{}
+	for _, serviceWorkloadLabels := range ug.RoleToWorkloadLabels[bindingSpec.Role] {
+		for key, value := range serviceWorkloadLabels {
+			workloadLabels[key] = value
+		}
+	}
+	if len(workloadLabels) > 0 {
+		authzPolicy.WorkloadSelector = &rbacproto.WorkloadSelector{Labels: workloadLabels}
+	}
+	authzPolicyConfig := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      model.AuthorizationPolicy.Type,
+			Group:     serviceRoleBinding.Group, // model.AuthorizationPolicy.Group is "rbac", but we need "rbac.istio.io".
+			Version:   model.AuthorizationPolicy.Version,
+			Name:      serviceRoleBinding.Name,
+			Namespace: serviceRoleBinding.Namespace,
+		},
+		Spec: &authzPolicy,
+	}
+	return authzPolicyConfig
+}
+
+// updateRoleToWorkloadLabels maps the ServiceRole name to workload labels that its rules originally
+// applies to.
+func (ug *Upgrader) updateRoleToWorkloadLabels(serviceRolePolicy model.Config) error {
+	roleName := serviceRolePolicy.Name
+	namespace := serviceRolePolicy.Namespace
+	serviceRoleSpec := serviceRolePolicy.Spec.(*rbacproto.ServiceRole)
+	for _, rule := range serviceRoleSpec.Rules {
+		for _, fullServiceName := range rule.Services {
+			if err := ug.mapRoleToWorkloadLabels(roleName, namespace, fullServiceName); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// mapRoleToWorkloadLabels maps the roleName from namespace to workload labels that its rules originally
+// applies to.
+func (ug *Upgrader) mapRoleToWorkloadLabels(roleName, namespace, fullServiceName string) error {
+	// TODO(pitlv2109): Handle when services = "*" or wildcards.
+	if strings.Contains(fullServiceName, "*") {
+		return fmt.Errorf("services with wildcard * are not supported")
+	}
+	serviceName := strings.Split(fullServiceName, ".")[0]
+	if _, found := ug.RoleToWorkloadLabels[roleName][serviceName]; found {
+		return nil
+	}
+	k8sService, err := ug.K8sClient.CoreV1().Services(namespace).Get(serviceName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if _, found := ug.RoleToWorkloadLabels[roleName]; !found {
+		ug.RoleToWorkloadLabels[roleName] = make(ServiceToWorkloadLabels)
+	}
+	if _, found := ug.RoleToWorkloadLabels[roleName][serviceName]; !found {
+		ug.RoleToWorkloadLabels[roleName][serviceName] = make(WorkloadLabels)
+	}
+	for key, value := range k8sService.Spec.Selector {
+		ug.RoleToWorkloadLabels[roleName][serviceName][key] = value
+	}
+	return nil
+}
+
+// checkAndCreateNewFile check if the name of the provided file is good, then create a new file with the
+// suffix.
+func checkAndCreateNewFile(fileName string) (*os.File, string, error) {
```

Comment:
You can just output to `stdout`, no need to output to a file. Customer could easily pipe the `stdout` to a file.

Reply:
Will change if you think this is a much better UX. Let's allow one file at a time for now.